Add support for descendant selectors.

xymostech opened the pull request:
Summary: This adds support for doing descendant style selection; that is,
changing child component styles based on the state/existence of a parent. This
selection in particularly useful when combined with the use of pseudo-classes,
by generating styles that look like `.parent:hover .child`, so the child styles
can depend on the active state of the parent.

Fixes #10

An example of the syntax that is added in this commit (and an example for the
explanation in this commit message):

```js
const styles = StyleSheet.create({
    parent: {
        '>>child': {
            color: "red"
        },
        ':hover': {
            '>>child': {
                color: "blue"
            },
            '>>otherchild': {
                color: "white"
            }
        },
    },

    altparent: {
        ':hover': {
            '>>child': {
                color: "green"
            }
        }
    }
});
```

It turns out that the API for aphrodite doesn't play well with descendent
styles. In particular, it's difficult to let *descendants have unique
classnames* for themselves, at the same time as allowing *merging between
styles which contain descendant selectors*.

This pull request attempts to do both of these things. The code is a bit messy,
so I'll lay out what's going on in this pull request. Please ask questions, if
something is confusing! The Aphrodite code is nice because it's all fairly
easily understood, and I'm afraid that this adding too much complexity.

The basic flow of this diff is:

1. In `StyleSheet.create`, we recurse through the passed-in styles, and find
each of the descendant selectors (which have keys that look like `>>blah`). In
the example above, it would find `parent['>>child']`,
`parent[':hover']['>>child']`, `parent[':hover']['>>otherchild']`, and
`altparent[':hover']['>>child']`. In each place, we:
  - generate a class name for that descendant selector. This is based on the
    class name of the parent class, as well as the key name.

    For example, if the class name for `styles.parent` was `parent_abcdef`, we
    might generate the class name `parent_abcdef__child` for `parent['>>child']`.

  - tag the style by adding a `_names` object, with the class name as a key of
    the object.

    For example, `parent['>>child']` would end up looking like `{ color: "red",
    _names: { parent_abcdef__child: true } }`.

  - collect a map of each of the keys (without the `>>` bit) to their class
    names.

    For example, for `styles.parent`, we would generate a map that looks like
    `{ child: "parent_abcdef__child", otherchild: "parent_abcdef__otherchild"
    }`.
We merge in the map from key to class name into the generated style, so that
the class names can be accessed using a syntax like `styles.parent.child`.

2. When *parent* styles are passed into `css()`, their styles are merged
together. If one style overrides another's descendant styling, the `_names`
object will be merged together and will contain all of the associated class
names.  
For example, when evaluating `css(styles.parent, styles.altparent)`, we would
end up with merged styles looking like:
    ```
    {
        '>>child': {
            color: "red",
            _names: { parent_abcdef__child: true },
        },
        ':hover': {
            '>>child': {
                color: "green",
                _names: {
                    parent_abcdef__child: true,
                    altparent_123456__child: true,
                },
            },
            '>>otherchild': {
                color: "white",
                _names: { parent_abcdef__otherchild: true },
            }
        }
    }
    ```
We then generate a map from the descendent keys to all of the class names that
could be associated with a given key by recursing and looking at each of the
`_names` objects. For example, the map would look like:
    ```
    {
        '>>child': ["parent_abcdef__child", "altparent_123456__child"],
        '>>otherchild': ["parent_abcdef__otherchild"]
    }
    ```
When generating the styles, we look at this map and then generate styles for
each of the classnames listed. This is so that these styles will match up with
uses of both `css(styles.parent.child)` and `css(styles.altparent.child)`.
For example, when generating the `style[':hover']['>>child']` styles, we
generate:
    ```
    .parent_abcdef-o_O-altparent_123456:hover .parent_abcdef__child { ... }
    .parent_abcdef-o_O-altparent_123456:hover .altparent_123456__child { ... }
    ```

3. When *descendant* styles are passed into `css()`, like
`css(styles.parent.child)`, we simply return the associated class name (in this
case, `"parent_abcdef__child"`) in the output.

Test Plan:
 - `npm run test`
 - `cd examples && npm run examples`, then visit http://localhost:4114/ and see
   that the last line starts green and when "Hover over me" is hovered, the
   other part turns blue.

@zgotsch @jlfwong @kentcdodds @montemishkin

kentcdodds commented on README.md:
````diff
@@ -173,6 +173,37 @@ const styles = StyleSheet.create({
+    }
+});
+```
+
````
Could we add the example css output of this?

xymostech replied:
Done!

kentcdodds commented on README.md:
```diff
@@ -173,6 +173,37 @@ const styles = StyleSheet.create({
+const styles = StyleSheet.create({
+    parent: {
+        ":hover": {
+            ">>my_child": {
```
personal preference: use `myChild` rather than `my_child` just because I don't like `snake_case` in my JavaScript (it's common to lint against it) and I'd prefer examples to use a style that's generally acceptable for JavaScript object property names.

xymostech replied:
Oops, my python was leaking. Done!

kentcdodds commented on README.md:
````diff
@@ -173,6 +173,37 @@ const styles = StyleSheet.create({
+Descendant rules (i.e. a rule like `.parent:hover .child { ... }` in CSS) are created by adding values to your styles indicating the styles to apply to the child. To distinguish the styles from normal CSS keys, the keys start with a `>>`, like `'>>child'`.
+
+```js
+const styles = StyleSheet.create({
````
This is great that we could accomplish this without having to have a special API for child selectors :+1:

kentcdodds commented on dist/aphrodite.js:
```diff
@@ -60,10 +60,48 @@ module.exports =
 
 	var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_ ...
 
+	var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };
```
I would expect the generated files to only be generated in a release commit. Helps keep the PRs scoped to the actual changes.

jlfwong replied:
Eugh, this should be collapsed as a generated file from setting the `.gitattributes`. I'm not sure why it's not.

kentcdodds commented on src/generate.js:
```diff
@@ -2,39 +2,89 @@ import Prefixer from 'inline-style-prefixer';
+        } else if (key[0] === '>' && key[1] === '>') {
+            // So we don't generate weird "_names: [Object object]" styles,
+            // make a copy of the styles and get rid of the _names value.
+            const stylesWithoutNames = {
```
You could avoid the use of `delete` (which has negative perf implications) by doing this:

```javascript
const {_names, ...stylesWithoutNames} = style[key]
```

[For example](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Creact%2Cstage-2&experimental=false&loose=false&spec=false&code=const%20style%20%3D%20%7B%0A%20%20bar%3A%20%7B%0A%20%20%20%20_names%3A%20'blah'%2C%0A%20%20%20%20otherProperty%3A%201%2C%0A%20%20%20%20otherProp%3A%202%2C%0A%20%20%7D%0A%7D%0A%0Aconst%20key%20%3D%20'bar'%0A%0A%2F%2F%20the%20relevant%20code%3A%0Aconst%20%7B_names%2C%20...stylesWithoutNames%7D%20%3D%20style%5Bkey%5D%0Aconsole.log(stylesWithoutNames))

:-)

xymostech replied:
:O that's so much cleaner!

xymostech replied:
Done!

kentcdodds commented:
Tried to review where I could. I'm afraid I still have a very limited understanding of how all this stuff works, but most stuff looks good to me :D

jlfwong commented on README.md:
````diff
@@ -173,6 +173,37 @@ const styles = StyleSheet.create({
+
+```js
+const styles = StyleSheet.create({
+    parent: {
````
Both in the original spec you described and in this documentation, it wasn't clear to me whether `parent` was just a style name or whether `parent` carries some special meaning. It seems to be the latter, but is worth clarifying and perhaps using a different name.

xymostech replied:
Done!

jlfwong commented on src/generate.js:
```diff
@@ -2,39 +2,89 @@ import Prefixer from 'inline-style-prefixer';
+    importantify,
 } from './util';
 
+// In a bundle of styles, find all of the different class names that a single
```
What's a "bundle of styles"?

xymostech replied:
Done! (and in a couple other places as well)

jlfwong commented on README.md:
```diff
@@ -173,6 +173,37 @@ const styles = StyleSheet.create({
 
+# Descendant selectors
+
+Descendant rules (i.e. a rule like `.parent:hover .child { ... }` in CSS) are created by adding values to your styles indicating the styles to apply to the child. To distinguish the styles from normal CSS keys, the keys start with a `>>`, like `'>>child'`.
```
On initial read of this, I assumed that `child` would be a literal class name and it took me a couple read throughs to figure out that `child` would be a classname generated by Aphrodite.

It might be worth explaining here that this is really only for cases that can't be expressed directly through others mean (which, I believe, is only styling children of pseudo-selector-modified parents).

jlfwong commented:
Thanks for giving this so much thought in order to implement missing functionality while not compromising on the philosophy of Aphrodite, and thanks to @kentcdodds for the quick review!

I'm probably going to need to take multiple passes at this, but I'll start with the helpfully thorough PR description.

Can you start out the PR description with what this is trying to accomplish, not with what makes it difficult? If I understand correctly, this is specifically to address the fact that children of pseudo selectors are not implementable without hacks in Aphrodite, right? Even starting off by saying it fixes #10 would've been helpful (I see that it's indeed present at the bottom of the description).

```
.parent_abcdef-o_O-altparent_123456:hover .parent_abcdef__child { ... }
.parent_abcdef-o_O-altparent_123456:hover .altparent_123456__child { ... }
```

The `...` in this case will be identical, right? If I'm understanding correctly, the hashes would end up being the same (the `abcdef` vs `123456` has me a little confused). If it's true that the hashes match, it might be better to jut throw out the name wholesale and avoid the need to generate duplicate styles, especially since we were tinkering with the idea of doing this on all styles to reduce bytes sent over the wire.

jlfwong commented on src/index.js:
```diff
@@ -5,14 +5,43 @@ import {
+            const name = makeClassName(key, val);
+
             return [key, {
-                // TODO(emily): Make a 'production' mode which doesn't prepend
```
This is still relevant, isn't it?

jlfwong commented on src/generate.js:
```diff
@@ -2,39 +2,89 @@ import Prefixer from 'inline-style-prefixer';
+        classNamesForDescendant);
+}
+
+const generateCSSInner = (selector, style, stringHandlers,
```
This is getting pretty complicated, and could also be greatly aided by a JSdoc, IMO.

jlfwong commented on tests/index_test.js:
```diff
@@ -118,6 +119,44 @@ describe('css', () => {
+        });
+
+        assert.equal(
+            css(sheet.blue, sheet.base.child),
```
What happens if you have `css(sheet.base.child, sheet.blue)`? I'm guessing behaviour is reasonable, but would be good to encode in a test.

xymostech replied:
done!

montemishkin commented:
> Thanks for giving this so much thought in order to implement missing functionality while not compromising on the philosophy of Aphrodite, and thanks to @kentcdodds for the quick review!

> I'm probably going to need to take multiple passes at this...

Agreed!

Thoughts from my first pass:

Given your approach, and taking into account the helpful notes from @jlfwong and @kentcdodds, the implementation looks good!

I'll try to find some time ASAP to both take a deeper dive into the current implementation and also consider the overall design / approach.

natew commented:
Just want to say this is a really great idea, especially for the `:hover>child` which is a really common pattern I've run into.

Why not just use `>` rather than `>>`?

kentcdodds commented:
> Why not just use > rather than >>?

There was discussion on that in the issue. Because `>` has meaning in CSS (direct child) and this would not follow that meaning (any descendant) we had to come up with something else.

xymostech commented:
Okay, I got around to cleaning up this pull request a bit. Not much changed, but I added a bunch more docs so hopefully that should clear up some confusion.

@jlfwong, some responses to your comments:
> Can you start out the PR description with what this is trying to accomplish, not with what makes it difficult? If I understand correctly, this is specifically to address the fact that children of pseudo selectors are not implementable without hacks in Aphrodite, right? Even starting off by saying it fixes #10 would've been helpful (I see that it's indeed present at the bottom of the description).

I added a better intro to the description, hopefully that should be more helpful about what's going on. Sorry, I was approaching this with all of the context in #10!

> ```
.parent_abcdef-o_O-altparent_123456:hover .parent_abcdef__child { ... }
.parent_abcdef-o_O-altparent_123456:hover .altparent_123456__child { ... }
````
The ... in this case will be identical, right? If I'm understanding correctly, the hashes would end up being the same (the abcdef vs 123456 has me a little confused). If it's true that the hashes match, it might be better to jut throw out the name wholesale and avoid the need to generate duplicate styles, especially since we were tinkering with the idea of doing this on all styles to reduce bytes sent over the wire.

You are correct, the `...` is identical. Actually now that I think about it, maybe I should be combining those with a comma (somehow forgot that CSS supports that!) The hashes would not be the same here, since they're just using the hashes we get from hashing the `parent` and `altparent` objects.

xymostech commented:
Okay, I pushed a change that uses commas to join things instead of duplicating styles, hopefully that's a little better.

I also ran a code coverage tool and found a case I hadn't tested. I'm gonna go back and add tests for the other things we aren't covering right now, maybe we could add a test that ensures aphrodite stays at 100% code coverage!

kentcdodds commented:
> maybe we could add a test that ensures aphrodite stays at 100% code coverage!

You may consider adding a `check-coverage` script and add that as a git hook using [ghooks](https://npmjs.com/package/ghooks) ([like this](https://egghead.io/lessons/javascript-how-to-write-a-javascript-library-adding-code-coverage-checking?series=how-to-write-an-open-source-javascript-library), though these day's I recommend [nyc](https://npmjs.com/package/nyc)). I'd recommend also using [opt-cli](https://npmjs.com/package/opt-cli) so new contributors can get involved without being blocked by coverage.

Just some thoughts :-)

jlfwong commented on src/generate.js:
```diff
@@ -2,39 +2,177 @@ import prefixAll from 'inline-style-prefix-all';
+    importantify, uniquify, mapObj,
 } from './util';
 
+/**
```
Awww yeaaahhh JSdoc  💖

jlfwong commented on src/generate.js:
```diff
@@ -2,39 +2,177 @@ import prefixAll from 'inline-style-prefix-all';
+ *
+ * In StyleSheet.create, we add a special `_names` object to each place where a
+ * descendant style is used containing the classname associated with that
+ * descenant. Here, we traverse the styles and make a map from the descendant
```
s/descenant/descendant

jlfwong commented on src/generate.js:
````diff
@@ -2,39 +2,177 @@ import prefixAll from 'inline-style-prefix-all';
+ * all end up as keys on the final merged object.)
+ *
+ * For example, given styles looking like:
+ * ```
````
Since this is a private API, and therefore unlikely to be nicely presented as HTML docs, I don't think we need markdown indicators. (If you do want to keep them, you're missing a closing triple backtick before the `@returns` below)

jlfwong commented on src/generate.js:
````diff
@@ -2,39 +2,177 @@ import prefixAll from 'inline-style-prefix-all';
+ *   }
+ * }, ...);
+ * ```
+ * will make 5 calls to `generateCSSRuleset`:
````
Super helpful example 🌟

kentcdodds commented:
FYI, this PR can no longer merge as it has conflicts.

jlfwong commented on src/generate.js:
````diff
@@ -55,6 +193,27 @@ const runStringHandlers = (declarations, stringHandlers) => {
+ *
+ * Example:
+ * ```
+ * generateCSSRuleset(".blah", { color: "red" });
````
Would be good to include an example that involves a pseudo selector so show that `selector` isn't necessarily just a classname selector

jlfwong commented on src/index.js:
```diff
@@ -5,14 +5,43 @@ import {
+                _className: name,
+            };
+
+            styles[key]._names = { [name]: true };
```
I'm getting a little confused here. Why is this a map if it only ever contains one key/value pair? Am I missing something here? I see above there's a traversal in `findNamesForDescendants` throughout the whole `styles` structure to find all the ancestors of a given `>>child` selector, but I don't see how it could have more than a single name in a single location here. This suspicion is supported by the fact that only a single assignment of `_names` takes place at `StyleSheet.create` time as far as I can tell.

The example above shows:

```
+ *   ">>child": {
 + *     _names: { "parent1_child_x": true, "parent2_child_y": true },
 + *     color: "red"
 + *   },
```

How did that `_names` with two keys happen given this code?

xymostech replied:
More than one name could appear here when they get merged together by a call to `css()`. So if you have like
```
const styles = StyleSheet.create({
    a: {
        '>>child': {
            color: 'red'
        }
    },

    b: {
        '>>child': {
            color: 'blue'
        }
    }
});

css(styles.a, styles.b);
```
then during the call to `css()` when we merge the `a` styles and `b` styles, the `_names` object will end up containing the classname from `a.>>child` and `b.>>child`. Does that make sense?

jlfwong replied:
That's how you wind up with more than one value in the array produced by `findNamesForDescendants`, but that doesn't yield multiple keys in the `_names` object, right?

jlfwong commented on src/index.js:
```diff
@@ -49,11 +80,23 @@ const css = (...styleDefinitions) => {
 
-    const className = validDefinitions.map(s => s._name).join("-o_O-");
+    // Filter out "plain class name" arguments, which just want us to add a
+    // classname to the end result, instead of generating styles.
```
I think it's probably helpful to say `this is the case, e.g., for descendent selectors where the relevant CSS is generated by the call to `css(styles.foo)`, not the call to `css(styles.foo.someChild)`.

jlfwong commented on tests/generate_test.js:
```diff
@@ -128,4 +128,120 @@ describe('generateCSS', () => {
+                  '.foo .foo_123456__blue{color:green !important;}');
+    });
+
+    it('handles multiples of the same descendant name', () => {
```
I'm not sure what "multiples" refers to here

jlfwong commented on tests/generate_test.js:
```diff
@@ -128,4 +128,120 @@ describe('generateCSS', () => {
+                  '.foo .foo__blue{color:blue !important;}');
+    });
+
+    it('handles merging and multiples for descenant styles', () => {
```
s/descenant/descedant

jlfwong commented:
Okay, after a second pass, I think I have a better understanding of what this is doing, but am still fuzzy on a few details, and am becoming a little uneasy about some behaviour (if I understand correctly how this works) that would be surprising to me.

Here are my major reservations about the API design:

## If you use a child selector without using the associated parent selector, nothing happens.

If I'm understanding the code correctly, if you have the following:

```
const styles = StyleSheet.create({
    foo: {
        '>>someChild': { color: 'red'; }
    }
});
```

And in your component, you have:

```
<div>
     <div className={css(styles.foo.someChild)}>Yo</div>
</div>
```

Then the div containing `Yo` will receive no styles, but WILL receive a class name. This is especially bad when combined with a silent failure from a typo, like so:

```
<div className={css(styles.fo)}>
     <div className={css(styles.foo.someChild)}>Yo</div>
</div>
```

## Because of multiple names, there are multiple ways of referencing the same child.

Having multiple ways of accomplishing the same thing generally acts as a red flag of API design to me, but perhaps it's not sensibly avoidable in this case...?

```
const styles = StyleSheet.create({
    foo: {
        '>>someChild': { color: 'red'; }
    },
    bar: {
        '>>someChild': { color: 'green'; }
    }
});
```

Then, if I'm understanding correctly, the following are equivalent from a user perspective and will generate the same CSS (though not HTML?):

```
<div className={css(styles.foo, styles.bar)}>
     <div className={css(styles.foo.someChild)}>
</div>
```

and

```
<div className={css(styles.foo, styles.bar)}>
     <div className={css(styles.bar.someChild)}>
</div>
```

and

```
<div className={css(styles.foo, styles.bar)}>
     <div className={css(styles.bar.someChild, styles.foo.someChild)}>
</div>
```

## The semantics of nested descendent selectors is confusing to me

...or possibly not implemented correctly? The code goes through all the trouble of identifying what the possibly parent selectors might be, but from the looks of it, it wasn't done recursively? If you have this:

```
const styles = StyleSheet.create({
    a1: {
        '>>b': { '>>c': { color: 'red'; } }
    },
    a2: {
        '>>b': { '>>c': { color: 'green'; } }
    }
});
```

Then what happens if you do this?

```
<div className={css(a1, a2)}>
     <div className={css(a1.b)}>
        <div className={css(a2.c})>
        </div>
    </div>
</div>
```

Overall, the semantics of any examples involving multiple instances of `>>someChild` aren't terribly clear to me (both in mergers and in pseudo selectors). Are there any valid use cases of allowing nested descendent selectors that aren't accomplishable in some other way? How much simpler does the design become if we restrict descendant selectors to only being immediate children of pseudo-selectors, and not allowing nesting of descendant selectors?

It's also not intuitive to me that `>>b` is addressed as `a1.b`, but `>>c` is not addressed as `a1.b.c`, though I guess it's okay because `a2` is not the same kind of thing as `b`?

Do others find the semantics from a user-standpoint confusing here? I'm particularly interested in @zgotsch thoughts, who may be able to draw from his experience in a real codebase that resulted in #10 being opened in the first place.

From a code complexity standpoint, this change seriously impedes my ability to reason intuitively about what the behavior of the code will be in all cases, but I didn't see any unnecessary complexity given this API design, so it may be necessary evil.

xymostech commented:
@jlfwong Thank you for the review of this API! I agree, it's kinda confusing. I'm also super sad that this change makes the code more difficult to understand, so I'd love to figure out an API that resulted in more intuitive code. I couldn't come up with a better way to do it myself, but maybe together we can brainstorm something? Gonna respond to your different points:

> If you use a child selector without using the associated parent selector, nothing happens.

That is correct. I'm not super worried about this one though, because it's the same as how normal CSS works, no? It does make the "these descendant selectors are *_special_*" more apparent though.

> Because of multiple names, there are multiple ways of referencing the same child.

Yeah, this is fugly. You are correct, those are all style-equivalent. I guess I chose this because it seems like it would lead to the least amount of things-not-working. If you have some parent selectors and some associated child selectors, then any combination should Just Work. It seemed simpler from an external API perspective, but leads to complex code.

> The semantics of nested descendent selectors is confusing to me

They're kinda confusing to me too. It wouldn't be too much work to make `a1.b.c` work, but it's not clear that that's any better. We could just disallow nested child selectors, but that might be more confusing? I guess if we threw an error that would be okay.

wrt. your example, I have no idea what happens there. Actually maybe that doesn't work any more now that I implemented the comma stuff. Let's go check!

------

I guess overall, I really like the API for simple cases but gets much more confusing with complex cases. Maybe something similar but simplified would be great? I'm not at all attached to this implementation, it just ended up having a really clean external API (as opposed to the `cssWithDescendants` stuff, which ended up being really hard to use).

iamrane commented:
Why not using the more sass syntax
'& .child': {
   padding: 10
}

xymostech commented:
@iamrane The reasons we're not landing this is because the API is too complex/confusing. Once that gets worked out, we could figure out what syntax we would like to use.

aesopwolf commented:
Any thoughts on using just a space as the child selector instead of two angle brackets `>>child`?

```js
' child': {
    color: "red"
},
```

marcusnielsen commented:
@aesopwolf I feel white spaces are dangerous due to the low visibility. Try find a bug where you missed a white space or had the wrong number or wrong kind. It's not fun. Same issue with O/0 and other ambiguous characters.

xymostech commented:
Okay, I'm going to close this PR since it's stagnated a bunch and I'm not super enthusiastic that this approach is going to work. All discussion should continue in #10.
